Add the `NoAccountAuthenticator` variant

junkil-park opened the pull request:
### Description
This PR updates the TypeScript SDK to support the simulation enhancement AIP: https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-92.md.

Key Changes:

Simulation API Update: Allows users to simulate transactions without providing public keys by updating simulateTransaction to accept `PublicKey | null` instead of `PublicKey`. If null is provided, `NoAccountAuthenticator` is used as an authenticator. This new authenticator variant is supported by the Aptos VM as per [this PR](https://github.com/aptos-labs/aptos-core/pull/13714).

Multisig V2 Example Update: The multisig v2 example (multisig_v2.ts) is updated to reflect the change in the multisig transaction simulation behavior. To pre-check the multisig payload before creation, an entry function payload simulation with the multisig account as the sender and 0x0 as the fee payer is used.

### Test Plan
This PR includes the following test scenarios:
* Simulate a transaction without checking the authentication key when the public key is not provided (i.e., public key is null).
* Simulate a fee payer transaction with fee payer as `0x0` and no public key of the fee payer provided.
* Fail to execute a transaction when `NoAccountAuthenticator` is used.

### Related Links
N.A.

0xmaayan commented on tests/e2e/transaction/transactionSubmission.test.ts:
```diff
@@ -828,4 +830,33 @@ describe("transaction submission", () => {
+      });
+      try {
+        const authenticator = new AccountAuthenticatorNoAccountAuthenticator();
+        const response = await submitTransaction({
```
please follow the other test flows on this file, i.e

```
const transaction = await aptos.transaction.build.simple({...})
const authenticator = aptos.transaction.sign({...})
const response = await aptos.transaction.submit.simple({...})
```

junkil-park replied:
This test case may not fit into the usual test flow. It attempts to execute a transaction with `NoAccountAuthenticator` without signing. This should fail because `NoAccountAuthenticator` can be used only in simulation, not in execution. This case intentionally tries to misuse a lower level API such as `submitTransaction`.

0xmaayan replied:
I understand we dont need to sign, but we should still use `await aptos.transaction.submit.simple({...})` to submit the transaction

0xmaayan commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -438,13 +439,20 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
 
 export function getAuthenticatorForSimulation(publicKey: PublicKey) {
+  // TODO: check this.
+  if (publicKey.toString() === (new Ed25519PublicKey(new Uint8Array(32))).toString()) {
```
what are you trying to check here? It feels error-prone.

junkil-park replied:
I updated this part so that `null` (rather than `000...0`) represents the absence of a public key.

junkil-park replied:
Update this.

0xmaayan reviewed (commented):
left initial comments. can we gat a PR description on what we are changing/adding here and why?

junkil-park commented:
> left initial comments. can we gat a PR description on what we are changing/adding here and why?

Added!

junkil-park commented on examples/typescript-esm/multisig_v2.ts:
```diff
@@ -131,6 +133,21 @@ const fundMultiSigAccount = async () => {
+      data,
+      withFeePayer: true,
+    });
+    rawTxn.feePayerAddress = AccountAddress.ZERO;
```
Got it. Removed. Thank you!

junkil-park commented on src/cli/localNode.ts:
```diff
@@ -39,8 +39,8 @@ export class LocalNode {
   start() {
-    const cliCommand = "npx";
-    const cliArgs = ["aptos", "node", "run-local-testnet", "--force-restart", "--assume-yes", "--with-indexer-api"];
+    const cliCommand = "aptos";
```
This is a temporary change to run a custom built node locally. I will revert it before merging this PR later. I added "TODO" as a comment not to forget.

0xmaayan commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -438,13 +439,19 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
 }
 
-export function getAuthenticatorForSimulation(publicKey: PublicKey) {
+export function getAuthenticatorForSimulation(publicKey: PublicKey | null) {
```
```suggestion
export function getAuthenticatorForSimulation(publicKey?: PublicKey) {
```

0xmaayan commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -438,13 +439,19 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
 
-export function getAuthenticatorForSimulation(publicKey: PublicKey) {
+export function getAuthenticatorForSimulation(publicKey: PublicKey | null) {
+  if (publicKey === null) {
```
```suggestion
  if (!publicKey) {
```

0xmaayan commented on src/transactions/types.ts:
```diff
@@ -278,15 +278,15 @@ export type InputSimulateTransactionData = {
    * For a single signer transaction
    */
-  signerPublicKey: PublicKey;
+  signerPublicKey: PublicKey | null;
```
```suggestion
  signerPublicKey?: PublicKey;
```

0xmaayan commented on src/transactions/types.ts:
```diff
@@ -278,15 +278,15 @@ export type InputSimulateTransactionData = {
    * For a fee payer or multi-agent transaction that requires additional signers in
    */
-  secondarySignersPublicKeys?: Array<PublicKey>;
+  secondarySignersPublicKeys?: Array<PublicKey | null>;
```
```suggestion
  secondarySignersPublicKeys?: Array<PublicKey>;
```

0xmaayan commented on src/transactions/types.ts:
```diff
@@ -278,15 +278,15 @@ export type InputSimulateTransactionData = {
    * For a fee payer transaction (aka Sponsored Transaction)
    */
-  feePayerPublicKey?: PublicKey;
+  feePayerPublicKey?: PublicKey | null;
```
```suggestion
  feePayerPublicKey?: PublicKey;
```

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +604,77 @@ describe("transaction simulation", () => {
+          functionArguments: [new U64(1), receiverAccounts[0].accountAddress],
+        },
+      });
+      const [response] = await simulateTransaction({
```
```suggestion
      const [response] = await aptos.transaction.simulate.simple({
```

junkil-park replied:
Done!

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +604,77 @@ describe("transaction simulation", () => {
+      },
+      withFeePayer: true,
+    });
+    rawTxn.feePayerAddress = AccountAddress.ZERO;
```
dont need it

junkil-park replied:
Removed it.

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +604,77 @@ describe("transaction simulation", () => {
+    rawTxn.feePayerAddress = AccountAddress.ZERO;
+
+    const [response] = await aptos.transaction.simulate.simple({
+      signerPublicKey: null,
+      transaction: rawTxn,
+      feePayerPublicKey: null,
```
```suggestion
      transaction: rawTxn,
```

junkil-park replied:
Updated it.

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +604,77 @@ describe("transaction simulation", () => {
+    rawTxn.feePayerAddress = feePayerAccount.accountAddress;
+
+    const [response] = await aptos.transaction.simulate.simple({
+      signerPublicKey: null,
+      transaction: rawTxn,
+      feePayerPublicKey: null,
```
```suggestion
      transaction: rawTxn,
```

junkil-park replied:
Done updating it accordingly.

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +604,77 @@ describe("transaction simulation", () => {
+    rawTxn.feePayerAddress = AccountAddress.FOUR;
+
+    const [response] = await aptos.transaction.simulate.simple({
+      signerPublicKey: null,
+      transaction: rawTxn,
+      feePayerPublicKey: null,
```
```suggestion
      transaction: rawTxn,
```

junkil-park replied:
Updated it.

0xmaayan reviewed (commented):
Did a first round
- Please fix the linting errors - can run `pnpm lint` on the sdk root folder
- When we have a property `publicKey?:PublicKey` it means the `publicKey` is of type `PublicKey` or `undefined` - please use this syntax
- Please update CHANGELOG
- Also, tests are failing

junkil-park commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -30,9 +30,9 @@ export class Simulate {
+    signerPublicKey: PublicKey | null;
     transaction: AnyRawTransaction;
-    feePayerPublicKey?: PublicKey;
+    feePayerPublicKey?: PublicKey | null;
```
I think we need `null` (or something equivalent) to cover the new use case: 
`feePayerPublicKey` can have:
(1) `undefined` if it is not a fee-payer transaction,
(2) `somePublicKey` if it is a fee-payer transaction, and the public/auth key will be checked on-chain, or
(3) `null` if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain.
Among these, (3) is the new use case.

0xmaayan replied:
I see. imo, it is a bit weird (in the SDK context and in the JS dev ecosystem) to differentiate between `null` and `undefined` and specifically ask the dev to write `feePayerPublicKey: null` as it means something critical.

Following this change - what would be the cases a dev still wants to NOT skip the public/auth key check? Are we (i.e Aptos) will be going with skipping the check from now on? wondering if we can combine the 1 and 3 use cases and just skip the check by default.

If we still want to keep the check, I would suggest maybe to introduce a new optional property for the dev to provide if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain, something like
```
skipFeePayerPublicKeyCheck?:boolean
```

junkil-park replied:
> Following this change - what would be the cases a dev still wants to NOT skip the public/auth key check? Are we (i.e Aptos) will be going with skipping the check from now on? wondering if we can combine the 1 and 3 use cases and just skip the check by default.

Using a simulation, a dev may want to validate the public key that they have. Otherwise, they have to actually submit a transaction. (1) is different from (2) and (3) in terms of transaction types (fee-payer txn vs. simple txn).

junkil-park replied:
> If we still want to keep the check, I would suggest maybe to introduce a new optional property for the dev to provide if it is a fee-payer transaction, and the public/auth key check will be skipped on-chain, something like
> skipFeePayerPublicKeyCheck?:boolean

This sounds good for the fee-payer case!

However, it seems not clear how to apply in the secondary signer case: https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#r1686804774

0xmaayan replied:
the secondary signer case makes sense to me - thanks for clarifying

junkil-park commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -51,10 +51,10 @@ export class Simulate {
     transaction: AnyRawTransaction;
-    secondarySignersPublicKeys: Array<PublicKey>;
-    feePayerPublicKey?: PublicKey;
+    secondarySignersPublicKeys: Array<PublicKey | null>;
```
I proposed `secondarySignersPublicKeys: Array<PublicKey | null>` because some secondary signers may provide public keys for auth key check, while others may not. For example, `secondarySignersPublicKeys` can be `[publicKey1, null, publicKey3, null]` where only first and third secondary signers provided public keys.

junkil-park replied:
Updated it to `secondarySignersPublicKeys: Array<PublicKey | undefined>;`

0xmaayan reviewed (commented):
Let's address the PR comments, clean up the PR (lint and format errors) - then can do another final round

junkil-park commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -30,9 +30,9 @@ export class Simulate {
   @ValidateFeePayerDataOnSimulation
   async simple(args: {
-    signerPublicKey: PublicKey;
+    signerPublicKey: PublicKey | null;
```
Updated it accordingly.

junkil-park commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -51,10 +51,10 @@ export class Simulate {
   @ValidateFeePayerDataOnSimulation
   async multiAgent(args: {
-    signerPublicKey: PublicKey;
+    signerPublicKey: PublicKey | null;
```
Updated it according to our discussion.

junkil-park commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -51,10 +51,10 @@ export class Simulate {
-    secondarySignersPublicKeys: Array<PublicKey>;
-    feePayerPublicKey?: PublicKey;
+    secondarySignersPublicKeys: Array<PublicKey | null>;
+    feePayerPublicKey?: PublicKey | null;
```
Updated it accordingly.

junkil-park commented:
> Did a first round
> 
> * Please fix the linting errors - can run `pnpm lint` on the sdk root folder
Some are unavoidable, so I added `eslint-disable-next-line` for those cases.

> * When we have a property `publicKey?:PublicKey` it means the `publicKey` is of type `PublicKey` or `undefined` - please use this syntax
Done accordingly.

> * Please update CHANGELOG
Updated.

> * Also, tests are failing
They passed locally. It requires a custom localnet.

0xmaayan commented on examples/typescript/local_node.ts:
```diff
@@ -107,6 +107,7 @@ async function stopLocalNode() {
     // Query localnet endpoint
     await fetch("http://localhost:8080");
   } catch (err: any) {
+    // eslint-disable-next-line no-console
```
🙇‍♀️

0xmaayan commented on src/api/transactionSubmission/simulate.ts:
```diff
@@ -41,19 +41,19 @@ export class Simulate {
+   * @param args.signerPublicKey optional. The signer public key
    * @param args.transaction An instance of a raw transaction
-   * @param args.secondarySignersPublicKeys An array of the secondary signers public keys
+   * @param args.secondarySignersPublicKeys An array of the secondary signers public keys, each of which can be optional
```
lets add some more info, something like
```suggestion
   * @param args.secondarySignersPublicKeys An array of the secondary signers public keys, each of which can be optional to support the public key skip checking
```

junkil-park replied:
Added some info

0xmaayan commented on src/transactions/authenticator/account.ts:
```diff
@@ -169,3 +171,18 @@ export class AccountAuthenticatorMultiKey extends AccountAuthenticator {
 }
+
+/**
+ * AccountAuthenticatorNoAccountAuthenticator for no account authenticator
```
could we elaborate when/why one would want to use it?

0xmaayan commented on src/transactions/types.ts:
```diff
@@ -278,11 +278,11 @@ export type InputSimulateTransactionData = {
-  signerPublicKey: PublicKey;
+  signerPublicKey?: PublicKey;
   /**
    * For a fee payer or multi-agent transaction that requires additional signers in
```
please add some more explanation on why/when a public key can be undefined

junkil-park replied:
Added some explanation.

0xmaayan commented on tests/e2e/transaction/transactionSubmission.test.ts:
```diff
@@ -828,4 +829,32 @@ describe("transaction submission", () => {
     });
   });
+
+  describe("transactions with no account authenticator", () => {
```
can we add more tests for all the different public key use cases? with a pk, without a pk, with array of pks (some defined some arent), etc

0xmaayan reviewed (commented):
whats the migration process? looks like we first need to have the feature on mainnet? or are we good with only on devnet/testnet?
Can we make sure tests are passing with the localnet on main branch before merging it in?

0xmaayan commented on tests/e2e/transaction/transactionSimulation.test.ts:
```diff
@@ -602,4 +602,69 @@ describe("transaction simulation", () => {
     });
   });
+
+  describe("simulations with no account authenticator", () => {
```
can we add more tests for all the different public key use cases? with a pk, without a pk, with array of pks (some defined some arent), etc

junkil-park replied:
I add the test cases for the multiagent transaction cases (for array of pks) : https://github.com/aptos-labs/aptos-ts-sdk/pull/467/files#diff-f97493b76e508e849a876ecb35492119c3ebf540140bda3c1a7b3313a32be2deR670-R708

0xmaayan commented on tests/e2e/transaction/transactionSubmission.test.ts:
```diff
@@ -828,4 +829,32 @@ describe("transaction submission", () => {
   });
+
+  describe("transactions with no account authenticator", () => {
+    test("single signer with script payload", async () => {
```
```suggestion
    test("it fails to submit a transaction when authenticator in undefined", async () => {
```

junkil-park replied:
Updated it.

0xmaayan commented on examples/typescript-esm/multisig_v2.ts:
```diff
@@ -141,19 +142,39 @@ const createMultiSigTransferTransaction = async () => {
-  });
-
-  console.log("simulateMultisigTx", simulateMultisigTx);
+  // The simulation enhancement feature is not enabled on the devnet yet.
```
can be removed. it is on devnet https://aptos-org.slack.com/archives/C05NLAKJM9U/p1726612173049729

junkil-park replied:
Done removing it.

0xmaayan commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -416,9 +423,15 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
-      getAuthenticatorForSimulation(publicKey),
-    );
+    if (secondarySignersPublicKeys) {
+      secondaryAccountAuthenticators = secondarySignersPublicKeys!.map((publicKey) =>
```
```suggestion
      secondaryAccountAuthenticators = secondarySignersPublicKeys.map((publicKey) =>
```

0xmaayan commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -416,9 +423,15 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
-    secondaryAccountAuthenticators = secondarySignersPublicKeys!.map((publicKey) =>
-      getAuthenticatorForSimulation(publicKey),
-    );
+    if (secondarySignersPublicKeys) {
```
repeated code, lets move it into a function

gregnazario commented:
I rebased and fixed the conflicts.

junkil-park commented:
> I rebased and fixed the conflicts.

Thank you! We’re now waiting on the security review and the testnet/mainnet release before merging this PR.

junkil-park commented on src/transactions/transactionBuilder/transactionBuilder.ts:
```diff
@@ -512,16 +514,16 @@ export function generateSignedTransactionForSimulation(args: InputSimulateTransa
+        getAuthenticatorForSimulation(publicKey),
       );
+    } else {
+      secondaryAccountAuthenticators = new Array(transaction.secondarySignerAddresses.length)
```
@0xmaayan , I updated the code, reducing the required step.
